### ISSUE TITLE
Unable to authorize through client credentials

### ISSUE THREAD

**Issue by praveenvaltix** [Azure/go-autorest#331]
We are working on integrating our product with Azure and we use Go as our language. I am running into an issue with authorizing successfully through "client credentials" method.

import (
    "log"

    "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-08-01/network"
    "github.com/Azure/go-autorest/autorest"
    "github.com/Azure/go-autorest/autorest/azure/auth"
)

// client credentials
credentialsConfig: auth.ClientCredentialsConfig{
            ClientID:     clientID,
            ClientSecret: clientSecret,
            TenantID:     tenantID,
            AADEndpoint:  "servicesvaltix.onmicrosoft.com",
            Resource:     "aztest",
        },

// Get azure authorizer
func (s *AzureService) GetAuthorizer() (autorest.Authorizer, error) {
    log.Printf("Get client credentials config")
    log.Printf("TODO REMOVE for privacy %+v", s)

    return s.credentialsConfig.Authorizer()
}

// Get azure vnet client
func (s *AzureService) GetVnetClient() (*network.VirtualNetworksClient, error) {
    vnetClient := network.NewVirtualNetworksClient(s.subscriptionID)
    authorizer, err := s.GetAuthorizer()
    if err != nil {
        log.Printf("Error while getting authorizer %v", err)
        return nil, err
    }
    vnetClient.Authorizer = authorizer
    return &vnetClient, nil
}

// list vpcs
func (s *AzureService) ListVPCs() ([]csputils.VPCInfo, error) {
    log.Printf("In list VNets")
    vnetClient, err := s.GetVnetClient()
    if err != nil {
        log.Printf("Error %v", err)
        return nil, err
    }

    resp, err := vnetClient.ListAll(context.TODO())
    if err != nil {
        log.Printf("Error while getting vpcs %v", err)
        return nil, err
    }

    log.Printf("Dump resp %+v", resp)

    return nil, nil
}

This is the error I see in the log. (for security reasons, I am obfuscating the secrets and the IDs).

--

2018/10/11 22:36:24 In list VNets
2018/10/11 22:36:24 Get client credentials config
2018/10/11 22:36:24 TODO REMOVE for privacy &{serviceType:client_credentials clientID:XXX clientSecret:XXXXX tenantID:XXX subscriptionID: credentialsConfig:{ClientID:XXX clientSecret:XXXXX TenantID:XXX AADEndpoint:XXX Resource:aztest}}
2018/10/11 22:36:24 Error while getting vpcs azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions//providers/Microsoft.Network/virtualNetworks?api-version=2018-08-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post /XXXX/oauth2/token?api-version=1.0: unsupported protocol scheme ""'
2018/10/11 22:36:24 Error while getting vpcs:  azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions//providers/Microsoft.Network/virtualNetworks?api-version=2018-08-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post /XXX/oauth2/token?api-version=1.0: unsupported protocol scheme ""'

--

The errors are not useful. 

I was also trying to see if az cli has an option to use client credentials vs the web login method by default. Coming from the AWS world, access keys are the easiest way to use and I am assuming that client credentials method is equivalent.

I appreciate all of your help in looking into this.

**Comment by vladbarosan** [comment]
@praveenvaltix it seems you are setting the AAD endpoint and the resource wrongly. Have a look at the constructor:  [NewClientCredentialsConfig](https://github.com/Azure/go-autorest/blob/master/autorest/azure/auth/auth.go#L287)

Also I would suggest using [NewAuthorizerFromEnvironment](https://github.com/Azure/go-autorest/blob/master/autorest/azure/auth/auth.go#L44)

**Comment by praveenvaltix** [comment]
@vladbarosan thanks for the response. I will take a look. FYI, we are building a multi tenant system, so in the same process we have to be able to connect to multiple different azure cloud subscriptions. Will that method work with  NewAuthorizerFromEnvironment ?

**Comment by praveenvaltix** [comment]
@vladbarosan yes thats correct. We basically get REST API requests with information on behalf of each of our unique customers who have different azure cloud subscriptions. For each such request, I have to invoke a Azure API to manage a azure resource. The azure credential information for our customers, we will be retrieving from a backend database. 

So to clarify, for our requirements, can we use NewClientCredentialsConfig?

**Comment by vladbarosan** [comment]
@praveenvaltix in that case NewClientCredentialsConfig might be your best shot. Note that the default Resource from the constructor is for ARM (as you can see [here](https://github.com/Azure/go-autorest/blob/master/autorest/azure/auth/auth.go#L305). If you you need other resources you will need to set it to one of the values in [here](https://github.com/Azure/go-autorest/blob/master/autorest/azure/environments.go#L62). 

Also note that if you are planning to support any other environment than the Public Cloud you will need to use the values from those Environments.

**Comment by praveenvaltix** [comment]
@vladbarosan thank you very much for your quick responses. Really appreciate it.

Yes, we are only focussed on Azure public cloud resources on behalf of our customers.

I made the change as you had suggested.

# new client credentials	
credentialsConfig: auth.NewClientCredentialsConfig(
			clientID,
			clientSecret,
			tenantID,
),

The error I see now is below - (removing IDs and secrets for privacy reasons).

2018/10/12 15:38:21 In list VNets
2018/10/12 15:38:21 Get client credentials config
2018/10/12 15:38:21 TODO REMOVE for privacy &{serviceType:client_credentials clientID:XX clientSecret:XX tenantID:XX subscriptionID: credentialsConfig:{ClientID:XX ClientSecret:XX TenantID:XX AADEndpoint:https://login.microsoftonline.com/ Resource:https://management.azure.com/}}
2018/10/12 15:38:21 Error while getting vpcs network.VirtualNetworksClient#ListAll: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="SubscriptionNotFound" Message="The subscription 'providers' could not be found."
2018/10/12 15:38:21 Error while getting vpcs:  network.VirtualNetworksClient#ListAll: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="SubscriptionNotFound" Message="The subscription 'providers' could not be found."

As I can see, there are 3 variables here. My understanding is below -

Client ID - Application ID. 
Client Secret - Application secret.
Tenant ID - its the same as Directory ID of the customer.

Basically, I created a new Azure active directory application and the Client ID I am setting is from the "Application ID' of the new application. Client Secret is the secret I saved during registration.

Please let me know if I have made an incorrect assumption here. 

Thank you.

**Comment by vladbarosan** [comment]
@praveenvaltix  the auth part looks fine. The error says that the subscription "providers" is not found. Can you pass the subscriptionid to the vnetClient and make sure the sub is in that tenant and the created SP has access to it.

**Comment by praveenvaltix** [comment]
@vladbarosan that is fantastic. I missed setting the correct subscription ID in the vnet client and everything working smoothly now.

Thank you very much for all your help. I am closing the issue.
